feat: add syntax for specifying function type environments

alexvitkov opened the pull request:
# Description


## Problem\*

Work towards https://github.com/noir-lang/noir/issues/2334


## Summary\*

In https://github.com/noir-lang/noir/pull/2335  we allowed higher order with closure arguments to pass type-checking -- this PR implements the monomorphizer changes needed to make those calls work.

### overridden_types

A new property `overridden_types` is added in the Monomorphizer. It allows us to change the type of an expression, for the currently monomorphized function instance only. It's used for adding capture environments to function types.

### propagate.rs

A function `propagate_overridden_types_expr` is added in a new file - propagate.rs. After we override the type of a parameter to add a capture group, we need to walk the body of the function and also update the types of all expressions where it is used.

## Documentation

- [ ] This PR requires documentation updates when merged.

  <!-- If checked, check one of the following: -->

  - [ ] I will submit a noir-lang/docs PR.

  <!-- Submit a PR on https://github.com/noir-lang/docs. Thank you! -->

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  <!-- List / highlight what should be documented. -->
  <!-- Dev Rel will reach out for clarifications when needed. Thank you! -->

## Additional Context

<!-- Supplement further information if applicable. -->

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

TomAFrench commented on crates/nargo_cli/tests/execution_success/closures_passed_to_higher_order_fns/src/main.nr:
This should go in `compile_success_empty`

jfecher reviewed (changes requested):
I'm not confidant this approach with `overridden_types` and manually trying to propagate modified types during monomorphization will be a good base to build upon for subsequent features, even if we want to replace it eventually.

It looks like we can only pass in closures directly to functions as arguments in this PR (e.g. passing them in a tuple does not work). Although the later does issue a type error instead of panicking.

I think exposing the environment type parameter to users would be less of a change and allow us to implement functions accepting closures, even nested within other types, in addition to functions returning closures in one go.

alexvitkov commented:
I understsand, I'll add support for the syntax you mentioned in the issue - `fn[env](params) -> ret`.

alexvitkov commented:
I've made the following changes:
- Reverted all the changes relating to overridden_types
- Reverted https://github.com/noir-lang/noir/pull/2335, as letting calls to higher order fns with closures pass typechecking doesn't make sense anymore
- Implemented the syntax `fn[envtype1, envtype2, ...](param1, ...) -> ret` and added tests

jfecher reviewed (approved):
Looks good. Let's add the environment generics change as a separate PR. That should get us all the way towards releasing closures as a feature in Noir.

There's a minor style suggestion but it's not a blocker for this PR.
